pkg/controller/topology.go

```diff
@@ -461,8 +508,69 @@ func getTopologyKeys(csiNode *storagev1.CSINode, driverName string) []string {
 	return nil
 }
 
+func getTopologyFromNodeName(nodeName string, topologyKeys []string, nodeLister corelisters.NodeLister, kubeClient kubernetes.Interface, pvcKeyForStore string, pvcNodeStore *InMemoryStore) (term topologyTerm, selectedNodeLabels map[string]string, isMissingKey bool) {
+	term = make(topologyTerm, 0, len(topologyKeys))
+	// Get Node Here
+	var nodeLabels map[string]string
+	if nodeLister != nil {
+		node, err := nodeLister.Get(nodeName)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				// Read from the cache
+				nodeLabels, err = getNodeLabelsFromCache(pvcNodeStore, pvcKeyForStore, nodeName)
+				if err != nil {
+					klog.Warningf("no node labels for node %s found in memory cache", nodeName)
+					return nil, nil, true
+				}
+				for _, key := range topologyKeys {
+					v, ok := nodeLabels[key]
+					if !ok {
+						return nil, nil, true
+					}
+					term = append(term, topologySegment{key, v})
+				}
+				term.sort()
+				return term, nodeLabels, false
+			}
+		}
+		if node != nil {
+			// Add or update to cache for node
+			pvcNodeStore.UpdateNodeLabels(pvcKeyForStore, node.Labels)
+
+			for _, key := range topologyKeys {
+				v, ok := node.Labels[key]
+				if !ok {
+					return nil, nil, true
+				}
+				term = append(term, topologySegment{key, v})
+			}
+			term.sort()
+			return term, node.Labels, false
+		}
+	} else {
+		// nodeLister is nil, read from the cache directly
+		var err error
+		nodeLabels, err = getNodeLabelsFromCache(pvcNodeStore, pvcKeyForStore, nodeName)
+		if err != nil {
+			klog.Warningf("no node labels for node %s found in memory cache", nodeName)
+			return nil, nil, true
+		}
+		for _, key := range topologyKeys {
```

Comment:
This logic looks repeated 3 times in this function. It would be better to share it across all 3 branches.